---
title: "hcocena_markdown_V2_satellite"
author: "Marie Oestreich"
date: "26 2 2021"
output: html_document
---

# Suggest Top Most Variant Genes

  This function calculates the inflection points in the ranked variances to filter for the top most variant genes in a data-driven way.
  In this case, there are 5 inflection points per dataset and we will keep the highest one in both caases to remove as few genes as possible.

```{r suggest top most variant genes}

hcocena::suggest_topvar()

```


# Checking data distribution

  To get a better feeling for the datasets and to detect possible outliers or prominent differences within datasets, the distribution of count values across all samples per dataset can be visualized. Here we choose to visualize them as boxplots (alternative option is frequency distributions), we set log_2 = FALSE, because our data is already logged and we don't wish to log the scale again for plotting. And since we don't only want the plots to be saved as PDFs, but also have them printed here into the markdown, we set plot = TRUE.


```{r data distribution plots, warning = FALSE}

plot_sample_distributions(plot_type = "boxplot",
                          log_2 = F, 
                          plot = T)

```


# PCA 

  We will also plot a PCA of our data: One PCA will be created per dataset, we want to use all genes (other options would be 'topvar' (only top-most-variant genes) or 'network' (only genes present in the co-expression network)). For details on these options, see ?hcocena::PCA.
  We want to colour the samples by our variable of interest (which in both cases was an annotation column called 'merged'). You may define a user-defined color vector (optional).
    

```{r PCA, warning = FALSE}

PCA(which = "all", 
    color_by = c("merged", "merged"), 
    cols = c("IL4_array" = "#00A1D5FF", "IFNg_array" = "#79AF97FF", "baseline_array" = "#374E55FF",
             "IL4_seq" = "#00A1D5FF", "IFNg_seq" = "#79AF97FF", "baseline_seq" = "#374E55FF"))

```



# Cytoscape

  Model layout in Cytoscape (https://cytoscape.org/download.html). Cytoscape MUST BE OPEN for this to work. **To shorten the time needed, please read the docs to the export_to_cytoscape function (?export_to_cytoscape).** 
  
  NOTE: instead of re-running this yourself (might take some time to load), you can also open the Cytoscape session that we created throughout the run and inspect the network. That will be faster and will still allow you to import the network coordinates from Cytoscape back into R (you can then immediately jump to 'import_layout_from_cytoscape' and can skip the export). You can find the Cytescape file of the network in the repository's 'showcase' folder, it is called 'cytoscape_session.cys'.
  
## Export Network

  This function exports the network to Cytoscape. There you can apply a variety of Cytoscape's great layout algorithms.
  
```{r Cytoscape export}

export_to_cytoscape()

```


## Import Layout

  After deciding on a layout that optimally visualises the network structure, this function imports the nodes' coordinates into R.

```{r Import layout}

import_layout_from_cytoscape()

```



## Plot Network
  
  After importing the coordinates, they can be used to plot the network. Since the Cytoscape coordinates are saved under hcobject[["integrated_output"]][["cluster_calc"]][["layout"]], we will provide that to the layout parameter.

```{r Plot network, fig.width=10, fig.height=7}

plot_integrated_network(layout = hcobject[["integrated_output"]][["cluster_calc"]][["layout"]])

```



# Cluster scores

  For every gene, the ratio of its edges to genes in the same cluster to its total number of edges is determined. The corresponding values are returned as a data frame and a box plot is generated showing the scores for each of the clusters. The data frame and the plot object will be saved under hcobject[["satellite_outputs"]][["cluster_scores"]].
  
```{r cluster scores}

get_cluster_scores()

```


# Hub gene detection

  Hub gene detection is available for selected clusters or the entire network. Hub genes are determined using a combined ranking based on weighted degree centrality, weighted closeness centrality and weighted betweenness centrality. A high quality version of the plots will be available in your save folder.
  
  As an example, we will look for hub genes in the cluster 'lightblue'. We want the top 10 hub candidates (based on the centrality hub score), we don't want to save the network with the hub genes highlighted, the network should not be plotted using a tree layout, we are only interested in hubs that are also transcription factors (if there are less than 10 TFs in the cluster, we will get less than 10 hub candidates) and we set plot to TRUE so that the cluster network with hubs highlighted will be plotted (but not saved since save = FALSE).
        
  Mean expression values of found hubs will be plotted as a heatmap per cluster, rows are scaled. 
A table of hubs per cluster will be exported as an Excel file. This table is also returned as output to the hcobject.
  

  
```{r Hub gene detection}

find_hubs(clusters = c("lightblue"),
          top = 10, 
          tree_layout = F, 
          TF_only = "TF",
          save = T, 
          plot = T)

```
  


# Visualize specific gene set
  
  Plots the mean expression values per condition ('variable of interest') for the given genes for each dataset as a heatmap. Values are scaled across rows.
  
  Here, we plot genes identified by Xue et al., 2014 (https://doi.org/10.1016/j.immuni.2014.01.006) as IL4-induced and IFNg-induced genes.
  

```{r visualize specific gene set}

visualize_gene_expression(genes = c("STAT6", "MAFG", "ETS2", "RUNX1", "TFE3", "HES6", "IRF4", "FOXQ1", 
                                    "CREB5", "SPN", "SOX8", "SIP1", "FOXD2", "GTF3C3", "BATF3", "EGR3", "FOS"), 
                          name = "Geneset_IL4")

visualize_gene_expression(genes = c("MAFA", "IRF9", "IRF7", "STAT2", "STAT1", "NFIX", "ELF4", "ETV7"), 
                          name = "Geneset_IFNg")

```
  


# Plotting network coloured by GFC

  For visualization of the GFC for every gene under the different observed groups (variable of interest), the network can be additionally re-plotted once for every group, with nodes being coloured according to their GFC value. This provides a more detailed resolution of the information acquired from the module heatmap.
  
  In this example, these networks show very nicely how the stimulus-specific patterns are preserved across technology, showing the suitability of the GFC approach for integration: The biological signal dominates, not the signal introduced by the sequencing technology.
 

```{r plot network with GFCs, fig.width=10, fig.height=7}

plot_GFC_network()

```
